keps/sig-network/3698-multi-network/README.md

```diff
@@ -0,0 +1,363 @@
+defining the new model in a backwards compatible way.
+* Integrate this API with the existing core objects.
+* Define “reference implementation” that can be used in Kubernetes CI.
+* Define conformance tests.
+
+
+### Non-Goals
+
+Define the CNI implementation.
+
+## Proposal
+
+### Personas
+
+#### Cluster operator
+
+This role will be responsible for defining and managing PodNetworks (see below)
+that properly describe the infrastructure available for the cluster, namespace
+and workloads. This persona can define which users can “attach” to a specific
+PodNetwork.
+
+#### Application developer
+
+**Application developer** is the consumer of PodNetwork via referencing them in
+their workloads. Application developers usually will not create or remove the
+PodNetwork on their own.
+
+### Terminology
+
+* **PodNetwork** - name of the API object that will represent a network in
+  Kubernetes cluster
+* **Cluster Default PodNetwork** - This is the PodNetwork attached to the Pod
+  when no additional networking configuration is provided in Pod spec.
+* **Primary PodNetwork** - This is the PodNetwork inside the Pod which interface
+  is used for the default gateway.
+
+
+### User Stories
+
+All user stories represent the type of use cases the multi-networking API should
+be able to support. References to technologies or exact products does not
+indicate that this API will directly support them. User stories can be
+overlapping in some areas. We want to document the concrete use cases from an
+end-user standpoint -- they may (and should!) boil down to a few common
+primitives listed in the below requirements section.
+
+
+#### Story #1
+As a Cloud Native Network Function (CNF) vendor my workloads deals with 2 types
+of network traffic: control plane and dataplane. For regulatory compliance I have
+to isolate these 2 traffic types. To achieve this I require 2 separate network
+interfaces inside my Pod. The underlying implementation (outside the Pod) has to
+ensure isolation on either Layer-2 or Layer-3.
+
+<p align="center">
+  <img src="mn-story-1.png?raw=true" alt="multi-network story 1 network L2 isolation"/>
+</p>
+
+#### Story #2
+As a Cloud Native Network Function (CNF) vendor I require a HW-based interface
+(e.g. SRIOV VF) to be provisioned to my workload Pod. I need to leverage that HW
+for performance purposes (high bandwidth, low latency), that my user-space
+application (e.g. DPDK-based) can use. The VF will not use the standard netdev
+kernel module. The Pod’s scheduling to the nodes should be based on hardware
+availability (e.g. devicePlugin or some other way). This interface might not
+support some Kubernetes functionality like e.g. Services, I am willing to give
+that up for performance.
+
+<p align="center">
+  <img src="mn-story-2.png?raw=true" alt="multi-network story 2 HW interface"/>
+</p>
+
+#### Story #3
+I have implemented my Kubernetes cluster networking using a virtual switch. In
```

Comment:
I don't think that is a good use case "I have this network feature and I want to use it in kubernetes", the story should be about why people will need to use this ... I can see this as ,"I want to be able to define network segmentation in the API"

Reply:
kind of the Story 4 too

Reply:
If anything this is similar to story #1 but with vSwitch. Agree that this is focused on implementation rather then user. Should we remove this?

Reply:
Up to you, I think that requirement is legit, but I don't feel is an user story, what I want to point out with my comments is that for something be part of Core it has to demonstrate strong user stories that can be easily generalizable, see the precedent with ClusterCIDR ... if this is going to follow the path of GatewayAPI and ClusterCIDR to build a strong community using CRDs , then this is not really important

Reply:
@danwinship my point is that the implementations details should not leak to the APIs (we've done with iptables and Services in some aspects so we know this is not ideal)
I'm arguing that the fact that you develop something  with virtual switches should not drive the design

Reply:
> I have implemented my Kubernetes cluster networking using a virtual switch

what means a virtual switch here, a virtual switch on the nodes, across the cluster ?

if this is about creating islands of pods, then why is not network policies suffice to cover this user story?